### PR TITLE
[simulate-sharing] muti-user simulator

### PR DESCRIPTION
**Pull request by cboddy**
This updates the fuzzer to:
* simulate two users that are followers/following of each other
* implement the grant/revoke read access simulation operations

It verifies correctly for ~10 operations, going to try an extended run shortly.

### CODE REVIEW

File: src/peergos/server/tests/simulation/NativeFileSystemImpl.java [hunk 1]
```diff
@@ -42,8 +45,8 @@ private void init() {
 //                , sharedRoot,
 //                peergosShare
         )) {
-            accessControl.add(path, user(), Permission.WRITE);
-            accessControl.add(path, user(), Permission.READ);
+//            accessControl.add(path, user(), Permission.WRITE);
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/NativeFileSystemImpl.java]: should these still be here?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 2]
```diff
@@ -31,96 +35,115 @@
...
     }
 
-    private void rmdir() {
-        Path path = getRandomExistingDirectory();
+    private void rmdir(String user) {
+        Path path = index.getRandomExistingDirectory(user);
         //rm -r
-
-        for (Path p : new ArrayList<>(simulatedDirectoryToFiles.keySet())) {
+        //TODO
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java]: is there still something to do here?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 3]
```diff
@@ -136,63 +159,172 @@ private int getNextFileLength() {
         return bytes;
     }
 
-    public Simulator(int opCount, Random  random, Supplier<Simulation> getNextSimulation, long meanFileLength, FileSystem referenceFileSystem, FileSystem testFileSystem) {
-        if (! referenceFileSystem.user().equals(testFileSystem.user()))
-            throw new IllegalStateException("Users must be the same");
 
-        simulatedDirectoryToFiles.put(Paths.get("/"+ referenceFileSystem.user()), new ArrayList<>());
+    private static class FileSystems {
+        private final List<Pair<FileSystem, FileSystem>> userFileSystems;
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java]: do the pairs signify friends?

Comment (ianopolous) [reply src/peergos/server/tests/simulation/Simulator.java]: got it, they are the peergos and native version of the same user. Is there a var name that would make this clear?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 4]
```diff
@@ -136,63 +159,172 @@ private int getNextFileLength() {
...
+        }
     }
 
     private void run(Simulation simulation) {
+        String nextUser = fileSystems.getNextUser();
+        run(simulation, nextUser);
+    }
+
+    private void run(Simulation simulation, String user) {
+        Supplier<String> otherUser = () -> fileSystems.getNextUser(user);
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java]: I imagine it's going to be super useful to have a single line log with op : user [path]

File: src/peergos/server/tests/simulation/Simulator.java [hunk 5]
```diff
@@ -212,61 +344,102 @@ public void run() {
 
         LOG.info("Running file-system verification");
         boolean isVerified = verify();
-        LOG.info("System verified =  "+ isVerified);
+        LOG.info("System verified =  " + isVerified);
+    }
+
+    private boolean verifySharingPermissions(String user, Path path) {
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java]: This seems to be verifying the sharer's view of the permissions, but not what the sharee can actually read?

File: src/peergos/server/tests/simulation/Simulator.java [hunk 6]
```diff
@@ -212,61 +344,102 @@ public void run() {
...
+                for (Path missing : expectedFilesAndFolders) {
+                    LOG.info("filesystem " + fs + " is missing  the path " + missing);
+                    isUserVerified = false;
+                }
+            }
+
+            // contents
+            for (Path path : expectedFilesForUser) {
+                boolean verifyContents = verifyContents(user, path);
+                boolean sharingPermissionsAreVerified = verifySharingPermissions(user, path);
```
Comment (ianopolous) [inline src/peergos/server/tests/simulation/Simulator.java]: do we also verify permissions for directories?

File: src/peergos/server/tests/simulation/NativeFileSystemImpl.java (hunk 1 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/NativeFileSystemImpl.java]: removed

File: src/peergos/server/tests/simulation/Simulator.java (hunk 3 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java]: done

File: src/peergos/server/tests/simulation/Simulator.java (hunk 6 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java]: not  yet - that is still `TODO`

File: src/peergos/server/tests/simulation/Simulator.java (hunk 5 above)
Comment (cboddy) [reply src/peergos/server/tests/simulation/Simulator.java]: That is the case - will update to actually read/write now.

**Pull request merged**